ENH: stats: Add 'alternative' and confidence interval to pearsonr.

WarrenWeckesser opened the pull request:
The confidence interval is computed using Fisher's z transform.

WarrenWeckesser commented:
An example of the enhancement:
```
In [14]: x = np.array([1, 2, 3, 4.5, 7, 8, 9])

In [15]: y = np.array([.1, 2, 2, 2.5, 2.8, 4, 4.1])

In [16]: result = pearsonr(x, y, alternative='two-sided', confidence_level=0.99)

In [17]: result.r
Out[17]: 0.9259022066726147

In [18]: result.pvalue
Out[18]: 0.0027574400089756517

In [19]: result.fishers_ci
Out[19]: ConfidenceInterval(low=0.3283407577873795, high=0.9941620142528566)
```

mdhaber commented on scipy/stats/stats.py:
```diff
@@ -3736,13 +3781,25 @@ def pearsonr(x, y):
+    alternative : {'two-sided', 'greater', 'less'}, optional
+        Indicates the alternative hypothesis. The default value is
+        'two-sided'.
+    confdence_level : float, optional
```
```suggestion
    confidence_level : float, optional
```

WarrenWeckesser replied:
Done.

mdhaber commented on scipy/stats/stats.py:
```diff
@@ -3736,13 +3781,25 @@ def pearsonr(x, y):
+        Indicates the alternative hypothesis. The default value is
+        'two-sided'.
+    confdence_level : float, optional
+        The confidence level for the calculation of the confidence
```
I misread this the first time. Would `The confidence level for the calculation of the correlation coefficient confidence interval` be any better? Just

WarrenWeckesser replied:
Done.

mdhaber commented on scipy/stats/stats.py:
```diff
@@ -3809,6 +3866,9 @@ def pearsonr(x, y):
     and -1.  Because abs(r') for any sample x' and y' with length 2 will
     be 1, the two-sided p-value for a sample of length 2 is always 1.
 
+    For backwards compatibility, the object that is returned behaves
```
```suggestion
    For backwards compatibility, the object that is returned also behaves
```

WarrenWeckesser replied:
Done.

mdhaber commented on scipy/stats/stats.py:
```diff
@@ -3712,7 +3750,14 @@ def __init__(self, msg=None):
+                                   ['fishers_ci'])
+
+
+def pearsonr(x, y, *, alternative='two-sided', confidence_level=0.95):
```
Why do we require that additional arguments be keyword only? Something I should start doing?

WarrenWeckesser replied:
For API development, this ensures that we can later add a new keyword parameter without requiring it to be the last parameter in the definition.  On an esthetic level, it forces users to write more readable code.

We haven't formally incorporated this into a recommend coding style, but for functions with more than two or three parameters, I think it would be a good idea.

mdhaber commented on scipy/stats/stats.py:
```diff
@@ -3692,6 +3693,43 @@ def f_oneway(*args, axis=0):
+    v = 1 / (n - 3)
+    s = np.sqrt(v)
+
+    if alternative == "two-sided":
```
Makes sense. Looks good.

mdhaber replied:
But maybe add to the notes how it's calculated? Probably should mention somewhere that it's an asymptotic approximation based on certain assumptions.

mdhaber commented on scipy/stats/stats.py:
```diff
@@ -3736,13 +3781,25 @@ def pearsonr(x, y):
         Input array.
+    alternative : {'two-sided', 'greater', 'less'}, optional
+        Indicates the alternative hypothesis. The default value is
+        'two-sided'.
```
I think we should be more explicit about what the alternatives mean, personally. There

mdhaber commented on scipy/stats/tests/test_stats.py:
```diff
@@ -375,16 +375,24 @@ def test_length_two_neg2(self):
-    def test_more_basic_examples(self):
+    # Expected values computed with R 3.6.2 cor.test, except for r and the
+    # p-value for alternative='two-sided', which were computed with mpmath.
+    @pytest.mark.parametrize('alternative, pval, rlow, rhigh',
```
Checked [here](https://www.tutorialspoint.com/execute_r_online.php):
```
x <- c(1, 2, 3, 4)
y <- c(0, 1, 0.5, 1)
cor.test(x, y,  method = "pearson", alternative = "two.sided")
```

WarrenWeckesser replied:
👍

WarrenWeckesser commented:
I made some significant changes in the latest update.  The return value of `pearsonr` is now an object that has the method `fishers_ci(confidence_level=0.95)`.  So instead of always computing the confidence interval and storing it as an attribute of the return value, it is only computed when the user asks for it.

I moved all the code related to `pearsonr` to a separate file, `_pearsonr.py`.

I moved this PR out of draft status so it is ready for review.

(When it comes time, let's squash all those commits before merging.)

WarrenWeckesser commented:
Squashed and rebased to fix a merge conflict.

WarrenWeckesser commented:
The one test failure on TravisCI was the unrelated recurring failure of the `test_symmetric_modes` ARPACK test.

tylerjereddy commented:
@WarrenWeckesser @mdhaber @rlucas7 Is a merge imminent here? I see no activity in the last 20 days or so, and this is an enhancement rather than bugfix, so I'll be tempted to bump the milestone pretty soon.

mdhaber commented:
I'll take another look today.

mdhaber commented on scipy/stats/_pearsonr.py:
```diff
@@ -0,0 +1,329 @@
+import warnings
```
For the sake of review, can you move this back into `stats`? After this is merged, I'd be happy to review a trivial PR that simply moves this into its own file.

mdhaber replied:
There have been improvements to the docstring that need to be merged here. I will do the surgery.

mdhaber commented on scipy/stats/_pearsonr.py:
```diff
@@ -0,0 +1,329 @@
+
+        The object has the following method:
+
+        * `fishers_ci(confidence_level=0.95)`
```
```suggestion
        fishers_ci(confidence_level=0.95)
```

mdhaber commented on scipy/stats/tests/test_stats.py:
```diff
@@ -375,16 +375,25 @@ def test_length_two_neg2(self):
-        assert_allclose(r, 0.674199862463242)
-        assert_allclose(p, 0.325800137536758)
+        result = stats.pearsonr(x, y, alternative=alternative)
+        assert_allclose(result.r, 0.674199862463242, rtol=1e-12)
```
```suggestion
        assert_allclose(result.r, 0.6741998624632421, rtol=1e-12)
```
Generated from code above.

mdhaber commented on scipy/stats/tests/test_stats.py:
```diff
@@ -375,16 +375,25 @@ def test_length_two_neg2(self):
         assert_equal(p, 1)
 
-    def test_more_basic_examples(self):
+    # Expected values computed with R 3.6.2 cor.test, except for r and the
+    # p-value for alternative='two-sided', which were computed with mpmath.
+    @pytest.mark.parametrize('alternative, pval, rlow, rhigh',
+                             [('two-sided',
+                               0.325800137536758, -0.814939, 0.992307),
+                              ('less',
+                               0.8370999, -1, 0.9856009),
+                              ('greater',
+                               0.1629001, -0.6785654, 1)])
```
```suggestion
    # Expected values computed with R 3.6.2 cor.test, e.g.
    # options(digits=16)
    # x <- c(1, 2, 3, 4)
    # y <- c(0, 1, 0.5, 1)
    # cor.test(x, y, method = "pearson", alternative = "g")
    # correlation coeffficient and p-value for alternative='two-sided' calculated with
    # mpmath agree to 16 digits.
    @pytest.mark.parametrize('alternative, pval, rlow, rhigh',
                             [('two-sided',
                               0.3258001375368, 0.3258001375368, 0.9923069752362500),
                              ('less',
                               0.8370999312316, -1, 0.985600937290653),
                              ('greater',
                               0.1629000687684, -0.6785654158217636, 1)])
```

mdhaber reviewed (commented):
Went ahead and reviewed as-is. I'll want to check the correlation coefficient math and double check the alternative p-value math, but looks correct.

mdhaber commented on scipy/stats/_pearsonr.py:
```diff
@@ -0,0 +1,329 @@
+                                    self.alternative)
+
+
+def pearsonr(x, y, *, alternative='two-sided'):
```
We're doing this pretty much all over now, so never mind.

mdhaber commented on scipy/stats/_pearsonr.py:
```diff
@@ -0,0 +1,329 @@
+    if alternative == 'two-sided':
+        prob = 2*special.btdtr(ab, ab, 0.5*(1 - abs(np.float64(r))))
+    elif alternative == 'less':
+        prob = 1 - special.btdtr(ab, ab, 0.5*(1 - abs(np.float64(r))))
```
Not going to sweat it. Leave as a potential improvement.

mdhaber commented on scipy/stats/_pearsonr.py:
```diff
@@ -0,0 +1,329 @@
+    def __init__(self, msg=None):
+        if msg is None:
+            msg = ("An input array is nearly constant; the computed "
+                   "correlation coefficent may be inaccurate.")
```
```suggestion
                   "correlation coefficient may be inaccurate.")
```

mdhaber commented on scipy/stats/_pearsonr.py:
```diff
@@ -0,0 +1,329 @@
+from scipy._lib._bunch import _make_tuple_bunch
+
+
+def _pearsonr_fishers_ci(r, n, confidence_level, alternative):
```
Looks good.

mdhaber commented:
Superseded by gh-15586.
